File: 2023/day4.jl
using DataStructures

struct Card
    id::Int128
    my_numbers::Set{Int128}
    winnig_numbers::Set{Int128} 
end

function parse_line(line::AbstractString)::Card
    info, nums = split(line, ": ")
    _, id = split(info)
    id = parse(Int128, strip(id))

    my_nums, winning_nums = split(nums, " | ")
    my_nums = map(x -> parse(Int128, x), split(strip(my_nums)))
    winning_nums = map(x -> parse(Int128, x), split(strip(winning_nums)))

    return Card(id, Set(my_nums), Set(winning_nums))
end

function parse_input(file_name::String)::Array{Card}
    file = open(file_name, "r")

    res = []
    for line in readlines(file)
        card = parse_line(strip(line))
        push!(res, card)
    end
    close(file)
    return res
end

function count_winning_numbers(card::Card)::Int128
    common  = intersect(card.my_numbers, card.winnig_numbers)
    return length(common)  
end

function worth(card::Card)::Int128
    n = count_winning_numbers(card)
    if n == 0
        return 0
    end
    return 2^(n-1)
end

function worth_sum(cards::Array{Card})::Int128
    sum(worth(card) for card in cards)
end

function get_copies(card::Card)::Array{Int128}
    n = count_winning_numbers(card)
    return (card.id+1):(card.id+n)
end

function count_cards(cards::Array{Card})::Int128
    q = Queue{Any}()
    n = length(cards)

    for card in cards
        enqueue!(q, card)
    end

    count::Int128 = 0

    while !(isempty(q))
        card = dequeue!(q)
        count += 1
        copies = get_copies(card)

        for copy in copies
            for card in cards
                if card.id == copy
                    enqueue!(q, Card(copy, card.my_numbers, card.winnig_numbers))
                end
            end
        end
    end
    return count
end

const test_file = "inputs/day4.test.txt"
const file_name = "inputs/day4.txt"
cards = parse_input(file_name)
#println(cards)
w = worth_sum(cards)
println(w)
c = count_cards(cards)
println(c)

File: 2023/day4.py
from dataclasses import dataclass
from collections import deque

@dataclass
class Card:
    id: int
    my_numbers: set[int]
    winning_numbers: set[int]

    def count_winning_cards(self) -> int:
        common =  self.my_numbers & self.winning_numbers
        return len(common)
    
    def worth(self) -> int:
        n = self.count_winning_cards()
        if n == 0:
            return 0
        return 2**(n-1)
    
    def get_copies(self):
        n = self.count_winning_cards()
        yield from range(self.id+1, self.id+1+n)
    
def parse_line(line: str) -> Card:
    info, nums = line.split(": ")
    _, id = info.split()
    id = int(id)

    my_nums, winning = nums.split(" | ")
    my_nums = set(map(int, my_nums.strip().split()))
    winning = set(map(int, winning.strip().split()))

    return Card(id, my_nums, winning)

def parse_input(file_name: str) -> dict[int, Card]:
    with open(file_name, "r") as file:
        res = {}
        for line in file:
            card = parse_line(line.strip())
            res[card.id] = card
    return res

def worth_sum(cards: dict[int, Card]) -> int:
    return sum(card.worth() for card in cards.values())

def count_cards(cards: dict[int, Card]) -> int:
    q = deque(cards.values())
    count = 0

    while q:
        card = q.popleft()
        count += 1

        for copy in card.get_copies():
            q.append(cards[copy])
    return count


if __name__ == "__main__":
    FILE_NAME = "inputs/day4.txt"
    cards = parse_input(FILE_NAME)

    print(worth_sum(cards))
    print(count_cards(cards))



    